File: sdk/javascript/Device.test.ts
import Device from './Device';

test('device.getKeyComponents()', () => {
  const device = new Device(
    'device1',
    'org1',
    'device1',
    'Device of Org1 User1',
    new Date('2021-12-12T17:34:00-05:00'),
  );

  expect(device.getKeyComponents()).toEqual([device.organizationId, device.id]);
});

test('device.toObject()', () => {
  const device = new Device(
    'device1',
    'org1',
    'device1',
    'Device of Org1 User1',
    new Date('2021-12-12T17:34:00-05:00'),
  );
  const obj = {
    id: 'device1',
    organizationId: 'org1',
    name: 'device1',
    description: 'Device of Org1 User1',
    lastUpdateTime: new Date('2021-12-12T17:34:00-05:00'),
  };

  expect(device.toObject()).toEqual(obj);
});

test('device.serialize()', () => {
  const device = new Device(
    'device1',
    'org1',
    'device1',
    'Device of Org1 User1',
    new Date('2021-12-12T17:34:00-05:00'),
  );
  const serialized =
    '{"id":"device1","organizationId":"org1","name":"device1",' +
    '"description":"Device of Org1 User1",' +
    '"lastUpdateTime":"2021-12-12T22:34:00.000Z"}';

  expect(device.serialize()).toEqual(serialized);
});

test('device.validate()', () => {
  const device = new Device('', '', '');

  expect(() => device.validate()).toThrow(/device ID/);
  device.id = 'device1';

  expect(() => device.validate()).toThrow(/organization ID/);
  device.organizationId = 'org1';

  expect(() => device.validate()).toThrow(/device name/);
  device.name = 'device1';

  expect(() => device.validate()).toThrow(/last update time/);
  device.lastUpdateTime = new Date('2021-12-12T17:34:00-05:00');

  expect(() => device.validate()).not.toThrow();
});

test('Device.fromObject()', () => {
  const obj = {
    id: 'device1',
    organizationId: 'org1',
    name: 'device1',
    description: 'Device of Org1 User1',
    lastUpdateTime: new Date('2021-12-12T17:34:00-05:00'),
  };

  const device = Device.fromObject(obj);
  expect(device.id).toEqual(obj.id);
  expect(device.organizationId).toEqual(obj.organizationId);
  expect(device.name).toEqual(obj.name);
  expect(device.description).toEqual(obj.description);
  expect(device.lastUpdateTime).toEqual(obj.lastUpdateTime);
});

test('Device.deserialize()', () => {
  const expected = new Device(
    'device1',
    'org1',
    'device1',
    'Device of Org1 User1',
    new Date('2021-12-12T17:34:00-05:00'),
  );
  const serialized =
    '{"id":"device1","organizationId":"org1","name":"device1",' +
    '"description":"Device of Org1 User1",' +
    '"lastUpdateTime":"2021-12-12T17:34:00-05:00"}';

  const actual = Device.deserialize(serialized);
  expect(actual.id).toEqual(expected.id);
  expect(actual.organizationId).toEqual(expected.organizationId);
  expect(actual.name).toEqual(expected.name);
  expect(actual.description).toEqual(expected.description);
  expect(actual.lastUpdateTime).toEqual(expected.lastUpdateTime);

  expect(() => Device.deserialize('\x00')).toThrow();
});

File: sdk/java/test/java/org/nexus_lab/iot_service_blockchain/sdk/DeviceTest.java
package org.nexus_lab.iot_service_blockchain.sdk;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;

import com.owlike.genson.JsonBindingException;
import java.time.OffsetDateTime;
import org.junit.Test;

public class DeviceTest {
  @Test
  public void testGetKeyComponents() {
    Device device = new Device();
    device.setId("device1");
    device.setOrganizationId("org1");
    device.setName("device1");
    device.setDescription("Device of Org1 User1");
    device.setLastUpdateTime(OffsetDateTime.parse("2021-12-12T17:34:00-05:00"));

    assertArrayEquals(
        device.getKeyComponents(), new String[] {device.getOrganizationId(), device.getId()});
  }

  @Test
  public void testSerialize() {
    Device device = new Device();
    device.setId("device1");
    device.setOrganizationId("org1");
    device.setName("device1");
    device.setDescription("Device of Org1 User1");
    device.setLastUpdateTime(OffsetDateTime.parse("2021-12-12T17:34:00-05:00"));

    String serialized =
        "{\"id\":\"device1\",\"organizationId\":\"org1\",\"name\":\"device1\","
            + "\"description\":\"Device of Org1 User1\","
            + "\"lastUpdateTime\":\"2021-12-12T17:34:00.000-05:00\"}";

    assertEquals(serialized, device.serialize());
  }

  @Test
  public void testValidate() {
    Device device = new Device();

    IllegalArgumentException exception =
        assertThrows(IllegalArgumentException.class, () -> device.validate());
    assertTrue(exception.getMessage().contains("device ID"));
    device.setId("device1");

    exception = assertThrows(IllegalArgumentException.class, () -> device.validate());
    assertTrue(exception.getMessage().contains("organization ID"));
    device.setOrganizationId("org1");

    exception = assertThrows(IllegalArgumentException.class, () -> device.validate());
    assertTrue(exception.getMessage().contains("device name"));
    device.setName("device1");

    exception = assertThrows(IllegalArgumentException.class, () -> device.validate());
    assertTrue(exception.getMessage().contains("last update time"));
    device.setLastUpdateTime(OffsetDateTime.parse("2021-12-12T17:34:00-05:00"));

    device.validate();
  }

  @Test
  public void testDeserialize() {
    Device expected = new Device();
    expected.setId("device1");
    expected.setOrganizationId("org1");
    expected.setName("device1");
    expected.setDescription("Device of Org1 User1");
    expected.setLastUpdateTime(OffsetDateTime.parse("2021-12-12T17:34:00-05:00"));

    String serialized =
        "{\"id\":\"device1\",\"organizationId\":\"org1\",\"name\":\"device1\","
            + "\"description\":\"Device of Org1 User1\","
            + "\"lastUpdateTime\":\"2021-12-12T17:34:00.000-05:00\"}";

    Device actual = Device.deserialize(serialized);
    assertEquals(expected, actual);

    assertThrows(JsonBindingException.class, () -> Device.deserialize("\u0000"));
  }
}
